cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -91,51 +201,628 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+     * with the new schedule.
+     */
+    public synchronized void updateScheduleTask(Id.Program program, SchedulableProgramType programType,
+                                                StreamSizeSchedule schedule)
+      throws ScheduleNotFoundException {
+      String scheduleId = AbstractSchedulerService.scheduleIdFor(program, programType, schedule.getName());
+      final StreamSizeScheduleTask scheduleTask = scheduleTasks.get(scheduleId);
+      if (scheduleTask == null) {
+        throw new ScheduleNotFoundException(schedule.getName());
+      }
+      scheduleTask.updateSchedule(schedule);
+
+      try {
+        StreamSize streamSize = pollOnce();
+        sendPollingInfoToActiveTasks(streamSize);
+      } catch (IOException e) {
+        LOG.debug("Ignoring stream events size polling after resuming schedule {} due to error",
+                  schedule.getName(), e);
+      }
+    }
+
+    /**
+     * Delete a scheduling task that is based on the data received by the stream referenced by {@code this} object.
+     */
+    public synchronized void deleteSchedule(Id.Program programId, SchedulableProgramType programType,
+                                            String scheduleName) throws ScheduleNotFoundException {
+      String scheduleId = AbstractSchedulerService.scheduleIdFor(programId, programType, scheduleName);
+      StreamSizeScheduleTask scheduleTask = scheduleTasks.remove(scheduleId);
+      if (scheduleTask == null) {
+        throw new ScheduleNotFoundException(scheduleName);
+      }
+      if (scheduleTask.isActive()) {
+        activeTasks.decrementAndGet();
+      }
+    }
+
+    /**
+     * Get the status a scheduling task that is based on the data received by the stream referenced by {@code this}
+     * object.
+     */
+    public ScheduleState scheduleTaskState(Id.Program programId, SchedulableProgramType programType,
+                                           String scheduleName) {
+      StreamSizeScheduleTask task = scheduleTasks.get(AbstractSchedulerService.scheduleIdFor(programId, programType,
+                                                                                             scheduleName));
+      if (task == null) {
+        return ScheduleState.NOT_FOUND;
+      }
+      return task.isActive() ? ScheduleState.SCHEDULED : ScheduleState.SUSPENDED;
+    }
+
+    public Id.Stream getStreamId() {
+      return streamId;
+    }
+
+    /**
+     * Poll the stream size using metrics.
+     */
+    private synchronized StreamSize pollOnce() throws IOException {
+      StreamSize streamSize = queryStreamEventsSize();
+      if (lastPollingInfo != null && streamSize.getSize() < lastPollingInfo.getSize()) {
+        for (StreamSizeScheduleTask streamSizeScheduleTask : scheduleTasks.values()) {
+          streamSizeScheduleTask.reset(streamSize.getTimestamp());
+        }
+      }
+      lastPollingInfo = streamSize;
+      delta = null;
```

Comment:
Oh, I see. I still think that it's valid to leave delta = null right here. Since when we only do polling, with no notification involved, the delta will not be reset and will also be wrong, unless metrics haven't increased since the last notification received.